lib/iris/common/metadata.py

```diff
@@ -1316,6 +1317,120 @@ def equal(self, other, lenient=None):
         return super().equal(other, lenient=lenient)
 
 
+class ConnectivityMetadata(BaseMetadata):
+    """
+    Metadata container for a :class:`~iris.coords.Connectivity`.
+
+    """
+
+    _members = ("cf_role", "start_index", "element_dim")
+
+    __slots__ = ()
+
+    @wraps(BaseMetadata.__eq__, assigned=("__doc__",), updated=())
+    @lenient_service
+    def __eq__(self, other):
+        return super().__eq__(other)
```

Comment:
I don't question that `element_dim` is a member of the `ConnectivityMetadata`, but I do question whether it should participate in equality.

For this we need to consider this in the context of the `Connectivity` class, with two instances, each of which have the transpose of the others `indices`, and each have the flip of the others `element_dim` - and let's say for arguments sake this is for the `face_node_connectivity` for a regular quad mesh with 3 faces.

So we have

```python
a_indices = array([[ 0,  1,  2,  3],
                   [ 4,  5,  6,  7],
                   [ 8,  9, 10, 11]])
a_element_dim = 1
```
and
```python
b_indices = array([[ 0,  4,  8],
                   [ 1,  5,  9],
                   [ 2,  6, 10],
                   [ 3,  7, 11]])
b_element_dim = 0
```
You could argue, given how we understand connectivity, these are actually equivalent e.g.,
```
>>> a_indices[0, :]
array([0, 1, 2, 3])

>>> b_indices[:, 0]
array([0, 1, 2, 3])
```
The `indices` have a different `shape`, but given our interpretation of how `element_dim` is used, the results of indexing is the same... and yet `element_dim` is different and `indices` are different.

If you agree to this interpretation (up for debate), then we need to support a different implementation of `__eq__` for `ConnectivityMetadata` and `Connectivity` :thinking:

Reply:
I would agree with this **if** the standard way of interacting with a `Connectivity` instance baked-in the intelligence around `src_dim` - i.e. a iterator that looks something like this:

```Python
for face_nodes in my_connectivity.src_iter():
   print(face_nodes)

> [0, 1, 2, 3]
> [4, 5, 6, 7]
```

Basically, we would need a recommended/safe/common way to iterate over two `Connectivity`s that are the transpose of each other. Without this, there would be dangers of it appearing safe to interact with / index two `Connectivity`s in the same way, even if that would have differing results.

Reply:
>If you agree to this interpretation (up for debate), then we need to support a different implementation of `__eq__` for `ConnectivityMetadata` and `Connectivity`

To be clear: `ConnectivityMetadata` equality does not care about `src_dim`, while `Connectivity` equality accounts for transposition by intelligently checking `indices` and `src_dim`. (if a user/developer really cares about true equality they can check `indices` and `src_dim` manually).

Reply:
116a683cd9eb310b4ea00cded8397c3e26d60d1c

Decided not to implement `__iter__`. Having thought at length, it seems too early to be make assumptions / be opinionated about how developers and advanced users will want to interact with this class. Instead I've explicitly created the [`indices_by_src()`](https://github.com/SciTools/iris/pull/3968/commits/116a683cd9eb310b4ea00cded8397c3e26d60d1c#diff-8fd786735e6d5fbb391db7380cb986b8836869167159414ca5a83141860b835dR3003-R3037) method, to make it maximally clear what's going on. If someone wants to iterate by `src_location` they can always iterate through the `indices_by_src()` output.

Reply:
@bjlittle what makes our discussion here regarding `src_dim` different from `start_index`?

As far as I can see, a `Connectivity` with 'shifted' `indices` and the alternate `start_index` is equivalent to a `Connectivity` without these, in exactly the same way as we've discussed a `Connectivity` with transposed `indices` and the alternate `src_dim`.

And it would be safe to treat these as equivalent if there was an agreed 'safe', 'neutralised' object that represents `indices` after correcting `start_index` differences, in the same way we discussed above for `src_dim` differences.

What have I missed? Is the problem that one couldn't offer up a **view** with shifted indices in the same way that one could with a transposed view, thus making such an operation more expensive?